pkg/backends/calico/client.go

```diff
@@ -174,16 +174,34 @@ func NewCalicoClient(confdConfig *config.Config) (*client, error) {
 		c.syncer.Start()
 	}
 
-	// Create and start route generator.
-	clusterCIDR := os.Getenv(envAdvertiseClusterIPs)
-	if len(clusterCIDR) > 0 {
-		if c.rg, err = NewRouteGenerator(c, clusterCIDR); err != nil {
+	// Get cluster CIDRs. Prefer the env var, if specified.
+	clusterCIDRs := []string{}
+	if clusterCIDR := os.Getenv(envAdvertiseClusterIPs); len(clusterCIDR) != 0 {
+		clusterCIDRs = []string{clusterCIDR}
+	} else if cfg != nil && cfg.Spec.ServiceExternalIPs != nil {
+		for _, c := range cfg.Spec.ServiceClusterIPs {
+			clusterCIDRs = append(clusterCIDRs, c.CIDR)
+		}
+	}
+
+	// Get external IP CIDRs.
+	externalCIDRs := []string{}
+	if cfg != nil && cfg.Spec.ServiceClusterIPs != nil {
+		for _, c := range cfg.Spec.ServiceExternalIPs {
+			externalCIDRs = append(externalCIDRs, c.CIDR)
+		}
+	}
+
+	if len(clusterCIDRs) != 0 || len(externalCIDRs) != 0 {
+		// Create and start route generator, if configured to do so. This can either be through
+		// environment variable, or the data store via BGPConfiguration.
+		if c.rg, err = NewRouteGenerator(c); err != nil {
 			log.WithError(err).Fatal("Failed to start route generator")
-		} else {
-			c.rg.Start()
 		}
+		c.rg.onClusterIPsUpdate(clusterCIDRs)
+		c.rg.onExternalIPsUpdate(externalCIDRs)
+		c.rg.Start()
```

Comment:
It feels like all this code for bootstrapping the route generator should be common with the cases where it gets started later on because of datamodel update.  Could that be arranged?

Reply:
By the way, I presume the reason for making this complex is so that confd can still run in a non-Kubernetes environment (including our own node ST).  Is that right?  If not, are there other reasons?

I took a slightly different tack for BGP passwords, and I guess it's not really ideal for us to follow different approaches for different features.  With BGP passwords, the overall approach is:
- Start the component (secretWatcher) unconditionally at start-of-day, but be tolerant of it failing to start up.  (Which implies not running in a Kubernetes environment.)
- If the component failed to start up, BGP passwords in the data model have no effect.

Whereas here the approach is:
- Start the component (routeGenerator) when environment or data model indicates that it's needed.
- Fail confd overall if the component is needed but fails to start up.

Let's discuss if we should do further work to align those better!